Readme.md

# last-error-middleware

  Middleware for catching the last `express.js` error and rendering it as json or an error page.

## Installation

    $ npm install last-error-middleware

## Example

```js
var lastError = require('last-error-middleware');

app.use(lastError({ pages: {
  '401': 'path/to/401.html',
  '404': 'path/to/404.html',
  '500': 'path/to/500.html',
  '503': 'path/to/503.html',
  '*'  : 'path/to/omg.html',
}));
```

## API

### .LastError(options)
  
  Construct a `last-error` middleware generator with custom `options`. The defaults are:

```js
{
  "pages": {},     // status code to html template path
  "stack": false   // return the error stack with json responses
}
```

### on('err', ..)

  Emitted on every error.

```js
var lastError = require('last-error-middleware');

var errors = lastError({ pages: {
  '401': 'path/to/401.html',
  '404': 'path/to/404.html',
  '500': 'path/to/500.html',
  '503': 'path/to/503.html'
});

app.use(errors);

errors.on('err', function (err) {
  console.log(err.stack);
});
```

## License

```
WWWWWW||WWWWWW
 W W W||W W W
      ||
    ( OO )__________
     /  |           \
    /o o|    MIT     \
    \___/||_||__||_|| *
         || ||  || ||
        _||_|| _||_||
       (__|__|(__|__|
```


History.md
0.3.0 - December 5, 2013
------------------------
 - Change to expose an express app

0.2.0 - December 3, 2013
-------------------------
 - Rename event `error` to `err`

0.1.2 - November 13, 2013
-------------------------
 - Adding `*` catch all error page

0.1.1 - October 29, 2013
-------------------------
 - Add emitter
 - Emit `error`

0.1.0 - October 29, 2013
-------------------------
 - Add 404 support
 - Change API to `errors.thrown()` and `errors.notFound`

0.0.6 - October 21, 2013
-------------------------
 - Only serves html if Accept header contains "html" and page is specified. More freely falls back to json.

0.0.5 - October 21, 2013
-------------------------
 - Optionally return the stack

0.0.4 - October 21, 2013
-------------------------
 - Removing `json:true` as an option
 - If application/json is in Accepts header, then json response is always served

0.0.3 - October 21, 2013
-------------------------
 - Removing `error.stack` from user response

0.0.2 - October 17, 2013
-------------------------
 - Checking header Access for `application/json`
 - Proper status code on html response

0.0.1 - October 17, 2013
-------------------------
:sparkles:


test/index.js

var assert = require('assert');
var error = require('http-error');
var express = require('express');
var html = require('hbs').__express;
var lastError = require('..');
var request = require('supertest');
var noop = function (){};


var err = error(401, 'UNAUTHORIZED', 'Not authorized');

function newApp () {
  var app = express()
    .use(express.bodyParser())
    .engine('html', html)
    .set('views', __dirname);

  app.all('/', function (req, res, next) {
    return next(err);
  });

  return app;
}

function checkJson (res) {
  var body = res.body;
  assert(body.error);
  assert(body.error.code === err.code);
  assert(body.error.status === err.status);
  assert(body.error.message === err.message);
}

describe('last-error', function () {
  it('should return 401 json error for xhr requests', function (done) {
    var app = newApp().use(lastError());
    request(app)
      .get('/')
      .set({ 'X-Requested-With': 'xmlhttprequest' })
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(err.status)
      .end(function (err, res) {
        if (err) return done(err);
        checkJson(res);
        done();
      });
  });

  it('should recognize application/json Accepts header', function (done) {
    var app = newApp().use(lastError());
    request(app)
      .get('/')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(err.status)
      .end(function (err, res) {
        if (err) return done(err);
        checkJson(res);
        done();
      });
  });

  it('should optionally return the stack', function (done) {
    var app = newApp().use(lastError({ stack: true }));
    request(app)
      .get('/')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(err.status)
      .end(function (err, res) {
        if (err) return done(err);
        checkJson(res);
        assert(res.body.error.stack);
        done();
      });
  });

  it('should render 401 page', function (done) {
    var app = newApp().use(lastError({ pages: { '401': __dirname + '/401.html' }}));
    request(app)
      .get('/')
      .set('Accept', 'text/html')
      .expect('Content-Type', /html/)
      .expect(200)
      .end(function (err, res) {
        if (err) return done(err);
        assert(res.text.indexOf('<h1>401</h1>') !== -1);
        done();
      });
  });

  it('should catch 404 errors', function (done) {
    var app = newApp().use(lastError());
    request(app)
      .get('/not/found')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(404)
      .end(function (err, res) {
        if (err) return done(err);
        var body = res.body;
        assert(body.error);
        assert(body.error.status === 404);
        assert(body.error.code === 'NOT_FOUND');
        done();
      });
  });

  it('should emit errors', function (done) {
    var app = newApp();
    var errors = lastError();
    app.use(errors);
    errors.on('err', function (err) {
      assert(err);
      assert(err.status === 404);
      assert(err.code === 'NOT_FOUND');
      done();
    });
    request(app)
      .get('/not/found')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(404)
      .end(noop);
  });
});

lib/index.js

var defaults = require('defaults');
var error = require('http-error');
var express = require('express');


/**
 * Expose `generate`.
 */

module.exports = generate;


/**
 * Return a new last error express app for mounting.
 *
 * @param {Object} options (optional)
 *   @property {Object} pages
 *   @property {Boolean} stack
 * @return {Express}
 */

function generate (options) {
  options = defaults(options || {}, {
    pages: {},
    stack: false
  });

  var app = express();

  /**
   * Last error middleware.
   *
   * @param {Error} err
   * @param {Request} req
   * @param {Response} res
   * @param {Function} next
   */

  function lastError (err, req, res, next) {
    app.emit('err', err);

    var header = req.get('Accept') || '';
    var pages = options.pages;
    var error = {
      code: err.code || 'INTERNAL',
      message: err.message || 'Internal error.',
      status: err.status || 500
    };

    if (options.stack) error.stack = err.stack;

    var page = pages[error.code] || pages[error.status] || pages['*'];
    if (page && header.indexOf('html') !== -1) {
      res.setHeader('Content-Type', 'text/html');
      res.render(page, { error: error });
    } else {
      res.setHeader('Content-Type', 'application/json');
      res.jsonp(error.status, { error: error });
    }
  }

  app.on('mount', function (parent) {
    parent
      .use(notFound)
      .use(lastError);
  });

  return app;
}


/**
 * Not found error middleware.
 *
 * @param {Request} req
 * @param {Response} res
 * @param {Function} next
 */

function notFound (req, res, next) {
  next(error(404, 'NOT_FOUND', 'Not found.'));
}